src/scene/materials/node-material.js

```diff
@@ -0,0 +1,692 @@
+                }
+                for (i = 0; i < this.graphData.subGraphs.length; i++) {
+                    if (this.graphData.subGraphs[i] && this.graphData.subGraphs[i] instanceof NodeMaterial && this.graphData.subGraphs[i].name) {
+                        if (this.graphData.subGraphs[i]._tmpRootTime === graphRootTime) {
+                            // recursion detected!
+                            return false;
+                        }
+
+                        this._tmpRootTime = graphRootTime;
+
+                        if (!(this.graphData.subGraphs[i].hasValidGraphData(graphRootTime))) {
+                            return false;
+                        }
+
+                    } else {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        }
+        return false;
+    },
+
+    // NB only called when shader graph is valid and shader is marked as dirty
+    initShader: function (device) {
+        // this is where we could get a list of pass types in current app render pipeline (layers)
+        // and query shader graph to check if there is a special output
+        // var passes=[SHADER_DEPTH, SHADER_FORWARD, SHADER_FORWARDHDR, SHADER_PICK, SHADER_SHADOW];
+        var passes = [SHADER_FORWARD];
+
+        for (var i = 0; i < passes.length; i++) {
+            var useSkin = false;
+
+            if (this.meshInstances && this.meshInstances[0] && this.meshInstances[0].skinInstance) {
+                useSkin = true; // TODO make this a variant?
+            }
+
+            var options = {
+                skin: useSkin,
+                shaderGraph: this,
+                pass: passes[i],
+                maxPixelLights: (this.maxPixelLights) ? this.maxPixelLights : 4,
+                maxVertexLights: (this.maxVertexLights) ? this.maxVertexLights : 8
+            };
+
+            // var depGraphTextureList=[];
+
+            var shaderDefinition = programlib.node.createShaderDefinition(device, options);
+            this.shaderVariants[passes[i]] = new Shader(device, shaderDefinition);
+        }
+    },
+
+    _addGraphVar: function (type, name, value) {
+        var graphVar;
+        if (value instanceof Texture) {
+            graphVar = { type: type, name: name, valueTex: value };
+        } else if (value instanceof Vec4) {
+            graphVar = { type: type, name: name, valueX: value.x, valueY: value.y, valueZ: value.z, valueW: value.w };
+        } else if (value instanceof Vec3) {
+            graphVar = { type: type, name: name, valueX: value.x, valueY: value.y, valueZ: value.z };
+        } else if (value instanceof Vec2) {
+            graphVar = { type: type, name: name, valueX: value.x, valueY: value.y };
+        } else if (typeof(value) === 'number') {
+            graphVar = { type: type, name: name, valueX: value };
+        } else {
+            graphVar = { type: type, name: name };

```

Comment:
I think perhaps it should actually fail to add a graph variable - since its not in the currently supported types - I will make this change. NB This shouldn't be a problem in Editor - because only supported types will be listed - but can be problem if creating graphs in code (but this is not supported yet).

Reply:
removed this bad case and added a comment